### PR TITLE
Add GATE and CATE for IRM models

### PR DESCRIPTION
**Pull request by SvenKlaassen**
### Description
Adding estimation of CATE and GATE for the IRM models based on the [Semenova et al.](https://academic.oup.com/ectj/article/24/2/264/5899048).

Implementation of a best linear predictor class [DoubleMLBLP](https://github.com/DoubleML/doubleml-for-py/blob/s-cate/doubleml/double_ml_blp.py), which estimates an ols model for a given orthogonal signal and basis. The estimation is done via [statsmodels.OLS](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html), which implements the correct variance estimation. 
The [DoubleMLBLP](https://github.com/DoubleML/doubleml-for-py/blob/s-cate/doubleml/double_ml_blp.py) includes a ``fit()`` method which estimates the coefficients and a ``confint()`` method, which creates the corresponding confidence intervals (and can be applied to a new basis vector to evaluate the best linear predictor on different values).

Additionally, the [DoubleMLIRM](https://github.com/DoubleML/doubleml-for-py/blob/s-cate/doubleml/double_ml_irm.py) class has been extended with the ``cate()`` and ``gate()`` methods. 
Both, ``cate()``  and ``gate()`` create a [DoubleMLIRMBLP](https://github.com/DoubleML/doubleml-for-py/blob/s-cate/doubleml/double_ml_blp.py) object by obtaining the orthogonal signal from the [DoubleMLIRM](https://github.com/DoubleML/doubleml-for-py/blob/s-cate/doubleml/double_ml_irm.py) model and fitting the linear projection for the supplied basis.  To obtain confidence intervals the ``confint()`` method has to be called. 
For the ``gate()`` method ``confint()`` without supplying new values for the basis outputs the confidence interval reduced to the different groups.

### Comments
Additional examples for the GATE and CATE estimation are available [here](https://github.com/DoubleML/doubleml-docs/tree/s-notebook-CATE/doc/examples) and can be added to the documentation (maybe the DGPs can be changed).

### PR Checklist

- [x] The title of the pull request summarizes the changes made.
- [x] The PR contains a detailed description of all changes and additions.
- [x] References to related issues or PRs are added.
- [x] The code passes all (unit) tests.
- [x] Enhancements or new feature are equipped with unit tests.
- [x] The changes adhere to the PEP8 standards.

### CODE REVIEW

File: doubleml/double_ml_blp.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+from scipy.linalg import sqrtm
+
+
+class DoubleMLBLP:
+    """Best linear predictor (BLP) for DoubleML with orthogonal signals.
+
+        Parameters
+        ----------
+        orth_signal : :class:`numpy.array`
+            The orthogonal signal to be predicted. Has to be of shape (n,).
```
Comment (PhilippBach) [inline doubleml/double_ml_blp.py:15]: I'd recommend to specify what `n` is here. I guess, users might wander if the `n` has to match some value, like the number of observations in the IRM model or so... Maybe it's good to briefly explain this in a sentence

File: doubleml/double_ml_blp.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        # blp of the orthogonal signal
+        g_hat = self._blp_model.predict(basis)
+
+        # calculate se for basis elements
+        blp_se = np.sqrt((basis.to_numpy().dot(self._blp_omega) * basis.to_numpy()).sum(axis=1))
+
+        if joint:
+            # calculate the maximum t-statistic with bootstrap
+            normal_samples = np.random.normal(size=[basis.shape[1], n_rep_boot])
+            bootstrap_samples = np.multiply(basis.to_numpy().dot(np.dot(sqrtm(self._blp_omega), normal_samples)).T,
```
Comment (PhilippBach) [inline doubleml/double_ml_blp.py:171]: see https://github.com/DoubleML/doubleml-for-py/pull/169/files#r1039459353

**Comment by PhilippBach** [comment]
Hi @SvenKlaassen , 

thanks for opening the PR and I think overall it looks really good. I believe, this is a very nice and useful feature. Except for my comments I posted in my review, there's a little (minor) thing I wonder whether it would make the usage a bit nicer, i.e.

* Add a method for string representation of the `DoubleMLBLP` class, i.e.,  `__str__()`  that might give users some insight on the ... I'm wondering which information users might be relevant for users who calculated the CATE, I had in mind
    * Is the grid used for constructing the confidence interval / bands the same as the one used for estimating the CATE (in-sample vs. out-of-sample)
    * Is it a CATE or a GATE estimator
    * Maybe some information on the model used for fitting, like score etc
    * $\ldots$ (we can discuss this together)

However, this is not crucial as it's not related to the major functionality of the new feature. I guess adding examples that explain proper usage will make things clear, too.

Btw I also dropped the `.coverage` file in https://github.com/DoubleML/doubleml-for-py/commit/9a18d0681a29cf7a9ddf5c4941f16ed3ece4b884

File: doubleml/double_ml_blp.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+class DoubleMLBLP:
+    """Best linear predictor (BLP) for DoubleML with orthogonal signals.
+
+        Parameters
+        ----------
+        orth_signal : :class:`numpy.array`
+            The orthogonal signal to be predicted. Has to be of shape (n,).
+
+        basis : :class:`pandas.DataFrame`
+            The basis for estimating the best linear predictor. Has to have the shape (n,d),
```
[the remark this answers is not in the document]
Comment (SvenKlaassen) [reply doubleml/double_ml_blp.py]: Thanks, I have changed this to `n_obs` and added more context.

File: doubleml/double_ml_blp.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        elif not (basis.shape[1] == self._basis.shape[1]):
+            raise ValueError('Invalid basis: DataFrame has to have the exact same number and ordering of columns.')
+        elif not list(basis.columns.values) == list(self._basis.columns.values):
+            raise ValueError('Invalid basis: DataFrame has to have the exact same number and ordering of columns.')
+
+        # blp of the orthogonal signal
+        g_hat = self._blp_model.predict(basis)
+
+        # calculate se for basis elements
+        blp_se = np.sqrt((basis.to_numpy().dot(self._blp_omega) * basis.to_numpy()).sum(axis=1))
```
[the remark this answers is not in the document]
Comment (SvenKlaassen) [reply doubleml/double_ml_blp.py]: Done,

**Comment by SvenKlaassen** [comment]
Hi @PhilippBach,

thanks for the review. I have incorporated the changes, extenden the documentation and added a summary.

**Comment by PhilippBach** [comment]
Hey @SvenKlaassen , 

I just updated the test jobs and merged them into this branch (➡️  https://github.com/DoubleML/doubleml-for-py/pull/169/commits/a7263a4c4f86e1ca7f27b96e405a95257c35401d) to let the jobs run

The changes look good to me, so in general, this PR is ready to be merged. If all checks are passed, you can merge it 👍

**Pull request merged**